File: HW1/hw1.py
import sys
from array import array

# -------- fast byte parser for non-negative ints --------
data = sys.stdin.buffer.read()
n = len(data)
i = 0

def next_int():
    global i
    # skip non-digits
    while i < n and data[i] <= 32:
        i += 1
    if i >= n:
        return None
    x = 0
    while i < n:
        c = data[i]
        if 48 <= c <= 57:  # '0'..'9'
            x = x * 10 + (c - 48)
            i += 1
        else:
            break
    # skip trailing sep (optional)
    while i < n and data[i] > 32:
        i += 1
    return x

# -------- read header --------
N = next_int()
K = next_int()
L = next_int()
S = next_int()

W = N + 1
SZ = W * W

# 2D prefix sums (row-major, 1D)
PF = array('I', [0]) * SZ  # forests == 1
PR = array('I', [0]) * SZ  # rocks   == 2

# helper to avoid attribute lookups in loops
PF_buf, PR_buf = PF, PR
Wloc = W

# -------- build prefix tables on the fly --------
for r in range(1, N + 1):
    base = r * Wloc
    base_up = (r - 1) * Wloc
    for c in range(1, N + 1):
        v = next_int()
        f = 1 if v == 1 else 0
        rck = 1 if v == 2 else 0
        idx = base + c
        PF_buf[idx] = f + PF_buf[base_up + c] + PF_buf[idx - 1] - PF_buf[base_up + c - 1]
        PR_buf[idx] = rck + PR_buf[base_up + c] + PR_buf[idx - 1] - PR_buf[base_up + c - 1]

# -------- rectangle sum via prefix table (1-based, inclusive) --------
def rect_sum(P, x1, y1, x2, y2):
    Wl = Wloc
    return (P[x2 * Wl + y2]
            - P[(x1 - 1) * Wl + y2]
            - P[x2 * Wl + (y1 - 1)]
            + P[(x1 - 1) * Wl + (y1 - 1)])

C = K - 2 * L  # central square size (>= 1)
best = 0

# -------- enumerate all KxK placements --------
for i_top in range(1, N - K + 2):
    i_bot = i_top + K - 1
    ic1 = i_top + L
    ic2 = ic1 + C - 1

    for j_left in range(1, N - K + 2):
        j_right = j_left + K - 1

        forests = rect_sum(PF, i_top, j_left, i_bot, j_right)

        jc1 = j_left + L
        jc2 = jc1 + C - 1
        rocks = rect_sum(PR, ic1, jc1, ic2, jc2)

        if rocks >= S and forests > best:
            best = forests

print(best)

File: HW1/hw1.c
#include <stdio.h>
#include <stdlib.h>

#define BUFSZ (1<<20)

static unsigned char ibuf[BUFSZ];
static size_t ipos = 0, ilen = 0;

static inline int refill() {
    ipos = 0;
    ilen = fread(ibuf, 1, BUFSZ, stdin);
    return (int)ilen;
}

static inline int next_int() {
    int c;
    // fetch first byte
    if (ipos >= ilen && !refill()) return 0;
    c = ibuf[ipos++];
    // skip non-digits (inputs here are non-negative)
    while ((c < '0' || c > '9')) {
        if (ipos >= ilen && !refill()) return 0;
        c = ibuf[ipos++];
    }
    int x = 0;
    while (c >= '0' && c <= '9') {
        x = x * 10 + (c - '0');
        if (ipos >= ilen && !refill()) break;
        c = ibuf[ipos++];
    }
    return x;
}

int main(void) {
    int N = next_int();
    int K = next_int();
    int L = next_int();
    int S = next_int();
    if (N <= 0) return 0;

    const int W = N + 1;
    const size_t SZ = (size_t)W * (size_t)W;

    // allocate and zero-initialise prefix tables
    int *PF = (int*)calloc(SZ, sizeof(int)); // forests (value == 1)
    int *PR = (int*)calloc(SZ, sizeof(int)); // rocks   (value == 2)
    if (!PF || !PR) {
        fprintf(stderr, "Allocation failed\n");
        free(PF); free(PR);
        return 1;
    }

    #define IDX(i,j) ((size_t)(i) * (size_t)W + (size_t)(j))

    // Build prefix sums while reading input
    for (int i = 1; i <= N; ++i) {
        for (int j = 1; j <= N; ++j) {
            int v = next_int();
            int f = (v == 1);
            int r = (v == 2);
            PF[IDX(i,j)] = f + PF[IDX(i-1,j)] + PF[IDX(i,j-1)] - PF[IDX(i-1,j-1)];
            PR[IDX(i,j)] = r + PR[IDX(i-1,j)] + PR[IDX(i,j-1)] - PR[IDX(i-1,j-1)];
        }
    }

    // helper: sum over [x1..x2] x [y1..y2], 1-based, inclusive
    #define RECTSUM(P,x1,y1,x2,y2) ( \
        (P)[IDX((x2),(y2))] - (P)[IDX((x1)-1,(y2))] - \
        (P)[IDX((x2),(y1)-1)] + (P)[IDX((x1)-1,(y1)-1)] )

    const int C = K - 2 * L; // central square size (>=1 guaranteed)
    int best = 0;

    // enumerate all KxK placements by top-left corner (i,j)
    for (int i = 1; i + K - 1 <= N; ++i) {
        int i2  = i + K - 1;
        int ic1 = i + L;
        int ic2 = ic1 + C - 1;
        for (int j = 1; j + K - 1 <= N; ++j) {
            int j2  = j + K - 1;

            // forests in whole KxK
            int forests = RECTSUM(PF, i, j, i2, j2);

            // rocks in central CxC (offset by L from each side)
            int jc1 = j + L;
            int jc2 = jc1 + C - 1;
            int rocks = RECTSUM(PR, ic1, jc1, ic2, jc2);

            if (rocks >= S && forests > best) best = forests;
        }
    }

    printf("%d\n", best);

    free(PF);
    free(PR);
    return 0;
}
